File: app/src/main/java/com/imadji/arch/sample/di/module/NetworkModule.java
package com.imadji.arch.sample.di.module;

import com.imadji.arch.data.source.remote.api.ApiService;
import com.imadji.arch.data.source.remote.api.TmdbApi;
import com.imadji.arch.data.source.remote.interceptor.RequestInterceptor;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import okhttp3.Interceptor;
import retrofit2.Retrofit;

/**
 * Created by imadji on 06/19/2019.
 */

@Module
public class NetworkModule {
    private String baseUrl;
    private String apiKey;

    public NetworkModule(String baseUrl, String apiKey) {
        this.baseUrl = baseUrl;
        this.apiKey = apiKey;
    }

    @Singleton
    @Provides
    List<Interceptor> provideInterceptors() {
        List<Interceptor> interceptors = new ArrayList<>();
        interceptors.add(new RequestInterceptor(apiKey));
        return interceptors;
    }

    @Singleton
    @Provides
    Retrofit provideRetrofit(List<Interceptor> interceptors) {
        return new ApiService(baseUrl).getRetrofit(interceptors);
    }

    @Singleton
    @Provides
    TmdbApi provideTmdbApi(Retrofit retrofit) {
        return retrofit.create(TmdbApi.class);
    }
}

File: app/src/main/java/com/imadji/arch/sample/di/module/NetworkModule.kt
package com.imadji.arch.sample.di.module

import com.imadji.arch.data.source.remote.api.TmdbRestClient
import com.imadji.arch.data.source.remote.api.MovieApi
import com.imadji.arch.data.source.remote.interceptor.RequestInterceptor
import com.imadji.arch.sample.BuildConfig
import java.util.ArrayList
import javax.inject.Singleton
import dagger.Module
import dagger.Provides
import okhttp3.Interceptor
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit

@Module
class NetworkModule(private val baseUrl: String, private val apiKey: String) {

    @Singleton
    @Provides
    fun provideInterceptors(): List<Interceptor> {
        val interceptors = ArrayList<Interceptor>()
        interceptors.add(createLoggingInterceptor())
        interceptors.add(RequestInterceptor(apiKey))
        return interceptors
    }

    @Singleton
    @Provides
    fun provideRetrofit(interceptors: MutableList<Interceptor>): Retrofit {
        return TmdbRestClient(baseUrl).getRetrofit(interceptors)
    }

    @Singleton
    @Provides
    fun provideMovieApi(retrofit: Retrofit): MovieApi {
        return retrofit.create(MovieApi::class.java)
    }

    private fun createLoggingInterceptor(): HttpLoggingInterceptor {
        val loggingInterceptor = HttpLoggingInterceptor()
        loggingInterceptor.level = if (BuildConfig.DEBUG) {
            HttpLoggingInterceptor.Level.BODY
        } else {
            HttpLoggingInterceptor.Level.NONE
        }
        return loggingInterceptor
    }
}